erizo/src/erizo/MediaStream.cpp

```diff
@@ -0,0 +1,515 @@
+}
+
+void MediaStream::sendPLIToFeedback() {
+  if (fb_sink_) {
+    fb_sink_->deliverFeedback(RtpUtils::createPLI(this->getVideoSinkSSRC(),
+      this->getVideoSourceSSRC()));
+  }
+}
+// changes the outgoing payload type for in the given data packet
+void MediaStream::sendPacketAsync(std::shared_ptr<DataPacket> packet) {
+  if (!sending_) {
+    return;
+  }
+  auto conn_ptr = shared_from_this();
+  if (packet->comp == -1) {
+    sending_ = false;
+    auto p = std::make_shared<DataPacket>();
+    p->comp = -1;
+    worker_->task([conn_ptr, p]{
+      conn_ptr->sendPacket(p);
+    });
+    return;
+  }
+
+  changeDeliverPayloadType(packet.get(), packet->type);
+  worker_->task([conn_ptr, packet]{
+    conn_ptr->sendPacket(packet);
+  });
+}
+
+void MediaStream::setSlideShowMode(bool state) {
+  ELOG_DEBUG("%s slideShowMode: %u", toLog(), state);
+  if (slide_show_mode_ == state) {
+    return;
+  }
+  asyncTask([state] (std::shared_ptr<MediaStream> connection) {
+    connection->stats_->getNode()[connection->getVideoSinkSSRC()].insertStat("erizoSlideShow", CumulativeStat{state});
+  });
+  slide_show_mode_ = state;
+  notifyUpdateToHandlers();
+}
+
+void MediaStream::muteStream(bool mute_video, bool mute_audio) {
+  asyncTask([mute_audio, mute_video] (std::shared_ptr<MediaStream> connection) {
+    ELOG_DEBUG("%s message: muteStream, mute_video: %u, mute_audio: %u", connection->toLog(), mute_video, mute_audio);
+    connection->audio_muted_ = mute_audio;
+    connection->video_muted_ = mute_video;
+    connection->stats_->getNode()[connection->getAudioSinkSSRC()].insertStat("erizoAudioMute",
+                                                                             CumulativeStat{mute_audio});
+    connection->stats_->getNode()[connection->getAudioSinkSSRC()].insertStat("erizoVideoMute",
+                                                                             CumulativeStat{mute_video});
+    connection->pipeline_->notifyUpdate();
+  });
+}
+
+void MediaStream::setVideoConstraints(int max_video_width, int max_video_height, int max_video_frame_rate) {
+  asyncTask([max_video_width, max_video_height, max_video_frame_rate] (std::shared_ptr<MediaStream> connection) {
+    connection->quality_manager_->setVideoConstraints(max_video_width, max_video_height, max_video_frame_rate);
+  });
+}
+
+void MediaStream::setFeedbackReports(bool will_send_fb, uint32_t target_bitrate) {
+  if (slide_show_mode_) {
+    target_bitrate = 0;
+  }
+
+  this->shouldSendFeedback_ = will_send_fb;
+  if (target_bitrate == 1) {
+    this->videoEnabled_ = false;
+  }
+  this->rateControl_ = target_bitrate;
+}
+
+void MediaStream::setMetadata(std::map<std::string, std::string> metadata) {
+  setLogContext(metadata);
+}
+
+WebRTCEvent MediaStream::getCurrentState() {
+  return connection_->getCurrentState();
+}
+
+void MediaStream::getJSONStats(std::function<void(std::string)> callback) {
+  asyncTask([callback] (std::shared_ptr<MediaStream> connection) {
```

Comment:
`connection` -> `stream`

Reply:
Agreed and done, now we get stats directly from the streams.